airflow/exceptions.py

```diff
@@ -256,12 +278,21 @@ def __init__(
         self.method_name = method_name
         self.kwargs = kwargs
         self.timeout = timeout
+
         # Check timeout type at runtime
         if self.timeout is not None and not hasattr(self.timeout, "total_seconds"):
             raise ValueError("Timeout value must be a timedelta")
 
+        # Check keyword arguments
+        if kwargs:
+            if method_name is None:
+                raise ValueError("Keyword arguments not allowed when method is not specified")
```

Comment:
Why not  just default `method_name: str = 'execute'` instead?

Reply:
So the only reason is to circumvent the behavior of adding "event" to the `kwargs` which ruins the logic for an operator that just implements the standard `execute` signature.

It makes it possible to use `pre_execute` as a general mechanism (or workaround) for adding logic before an operator runs – that's at least my example use-case here, delaying any odd operator without making a change.

Reply:
Question: is the only saving from this not putting `event=None` in your `execute()` method signature? Is that worth making a whole bunch of the API slightly more complex?

Reply:
The problem is that it's often times not my `execute()` method – lots of people use built-in operators or write their own ones that don't know about or want to know about deferred tasks.

Reply:
Recorrecting my thoughts here - it should be fine to have custom kwargs for `execute`, since it's something you can always override in an operator.

Reply:
> don't know about or want to know about deferred tasks

Wait, so you're injecting deferral into operators via the pre_execute hook that otherwise may not expect it? I would be wary of doing that, there might be side-effects from operators that do too much in their own init or pre-execute code...

Reply:
> It makes it possible to use pre_execute as a general mechanism (or workaround) for adding logic before an operator runs – that's at least my example use-case here, delaying any odd operator without making a change.

This sounds like a bit of an anti-pattern to me -- TimeDeltaSensor (+ an Async version) already exists, or change the schedule of the whole DAG to be N minutes later.

Reply:
On not allowing `kwargs` when `method_name` is unset, I just thought it was a little cleaner. I'm not sure I understand the part about custom kwargs for `execute` – how does that look/work?

Is this an anti-pattern? So I don't think so and there is some more discussion on this in #17576. For example, if I want to postpone a particular task (not the whole DAG, just a task) so that it runs no earlier than 7:30am then to me, that is a scheduling policy on the task, it's not helpful to have another "delayer" task in front of it because ultimately, that just contributes to noise.

I'm migrating some workflows from a different system where there is task-level scheduling that allows such configurations and it really works quite well. There is less clutter when each task has a little bit more "scheduling power". But of course, that's just a point of view.

Reply:
@andrewgodwin note that the operator should not set a `pre_execute` hook on itself, that would be weird I think due to the semantics around `pre_execute` – it runs sequentially just before `execute` so you could just put that logic into the `execute` method.

The idea with `pre_execute` is that a user can have some custom logic before the operator executes. They could subclass and come up with some name for the class, but this is a simple way to change the execution flow on an ad-hoc basis.